packages/eslint-plugin/src/rules/no-misused-object-likes.ts

```diff
@@ -0,0 +1,149 @@
+        additionalProperties: false,
+        properties: {
+          checkObjectKeysForMap: {
+            description: 'Check usage Object.keys for Map object',
+            type: 'boolean',
+          },
+          checkObjectValuesForMap: {
+            description: 'Check usage Object.values for Map object',
+            type: 'boolean',
+          },
+          checkObjectEntriesForMap: {
+            description: 'Check usage Object.entries for Map object',
+            type: 'boolean',
+          },
+          checkObjectKeysForSet: {
+            description: 'Check usage Object.keys for Set object',
+            type: 'boolean',
+          },
+          checkObjectValuesForSet: {
+            description: 'Check usage Object.values for Set object',
+            type: 'boolean',
+          },
+          checkObjectEntriesForSet: {
+            description: 'Check usage Object.entries for Set object',
+            type: 'boolean',
+          },
+        },
+      },
+    ],
+  },
+
+  create(context, [options]) {
+    const services = getParserServices(context);
+
+    function checkObjectMethodCall(
+      callExpression: TSESTree.CallExpression,
+    ): void {
+      const argument = callExpression.arguments[0];
+      const type = services.getTypeAtLocation(argument);
+      const argumentTypeName = type.getSymbol()?.name;
+      const callee = callExpression.callee as MemberExpression;
+      const objectMethod = (callee.property as Identifier).name;
+
+      if (argumentTypeName === 'Map') {
+        if (objectMethod === 'keys' && options.checkObjectKeysForMap) {
+          context.report({
+            node: callExpression,
+            messageId: 'objectKeysForMap',
+          });
+        }
+        if (objectMethod === 'values' && options.checkObjectValuesForMap) {
+          context.report({
+            node: callExpression,
+            messageId: 'objectValuesForMap',
+          });
+        }
+        if (objectMethod === 'entries' && options.checkObjectEntriesForMap) {
+          context.report({
+            node: callExpression,
+            messageId: 'objectEntriesForMap',
+          });
+        }
+      }
+      if (argumentTypeName === 'Set') {
+        if (objectMethod === 'keys' && options.checkObjectKeysForSet) {
+          context.report({
+            node: callExpression,
+            messageId: 'objectKeysForSet',
+          });
+        }
+        if (objectMethod === 'values' && options.checkObjectValuesForSet) {
+          context.report({
+            node: callExpression,
+            messageId: 'objectValuesForSet',
+          });
+        }
+        if (objectMethod === 'entries' && options.checkObjectEntriesForSet) {
+          context.report({
+            node: callExpression,
+            messageId: 'objectEntriesForSet',
+          });
+        }
+      }
+    }
+
+    return {
+      'CallExpression[callee.object.name=Object][callee.property.name=keys][arguments.length=1]':
```

Comment:
This is missing various other ones such as `Object.assign/groupBy/hasOwn?`, it might be better to move this check into the `checkObjectMethodCall` function and have a list of banned methods there.